plugins/inputs/mqtt_consumer/mqtt_consumer.go

```diff
@@ -362,17 +378,80 @@ func (m *MQTTConsumer) createOpts() (*mqtt.ClientOptions, error) {
 				server = "ssl://" + server
 			}
 		}
-
 		opts.AddBroker(server)
 	}
 	opts.SetAutoReconnect(false)
 	opts.SetKeepAlive(time.Second * 60)
 	opts.SetCleanSession(!m.PersistentSession)
 	opts.SetConnectionLostHandler(m.onConnectionLost)
-
 	return opts, nil
 }
 
+// parseMeasurement gets a single value from the incoming topic based on the user configuration (TopicParsing.Measurement)
+func parseMeasurement(keys []string, values []string) (string, error) {
+	for i, k := range keys {
+		if k != "_" {
+			return values[i], nil
+		}
+	}
+	return "", fmt.Errorf("no measurements found")
+}
+
+// parseTags gets multiple tags from the topic based on the user configuration (TopicParsing.Tags)
+func parseTags(keys []string, values []string) (map[string]string, error) {
+	results := make(map[string]string)
+	for i, k := range keys {
+		if k != "_" {
+			results[k] = values[i]
+		}
+	}
+	if len(results) == 0 {
+		return nil, fmt.Errorf("no tags found")
+	}
+	return results, nil
+}
+
+// parseFields gets multiple fields from the topic based on the user configuration (TopicParsing.Fields)
+func parseFields(keys []string, values []string, types map[string]string) (map[string]interface{}, error) {
```

Comment:
Maybe we should split this into extracting the values (which is the same as `parseTags`) and then do the type conversion in a second step? This way we can generalize the extracting, e.g. if adding more modifiers like `__` or whatever people come up with.

Reply:
These functions have been merged.